docs/src/contributing/pull-request.md

```diff
@@ -0,0 +1,135 @@
+   4. Now you can force-push the `fix_all_bugs ` branch to rewrite its history on the gitlabserver (necessary due to rebase that was made)
+5. Make sure you run this procedure before every commit that you intend to push for review:
+   1. Make sure your code builds without warnings - `./cargo build` is a good start
+   2. Run tests on all packages you have worked on ``` ./cargo test \--package \<crate\_name\> ```
+      1. Just running cargo test in the root directory will run a bunch of tests that will fail on your machine, only run the tests on the crates you touch.
+   3. If your crate has examples, they should get built by cargo test also, but they will not get run. Make sure they work correctly, CI will not hold your hand here.
+   4. ``` /scripts/cargo-fmt.sh && ./scripts/cargo-clippy.sh```
+   5. Make sure your Cargo.toml files are *perfect* [by sorting them](#unsorted-deps)
+   6. Make sure there are no trailing whitespaces anywhere `git diff origin/master --check --oneline`
+      1. If you find any in files that cargo fmt will not fix for you, you can use e.g. `sed  's/[ \t]*$//'` to fix them manually
+   7. You can manually run an approximation of full github CI with ``` ./ci/run-local.sh``` to diagnose any problems
+   8. Once all of the above checks pass, make the final commit and push `fix_all_bugs` branch to github
+      1. Make sure you [squash unnecessary commits](#pr-squash-commits) to ensure your commit log is sensible
+6. To create a PR:
+   1. Go to `ubercoder/agave` on github, switch to `fix_all_bugs` branch, and make PR against `anza-xyz/agave` master
+   2. Make sure you explain exactly why this PR is useful on its own merit
+   3. If you expect to follow with more PRs to expand on a feature, make it clear
+7. If you are not agave core developer, your PR will be reviewed by one, and, assuming it gets approved, CI process will be enabled for it.
+   1. If you are a developer, consider choosing “make draft PR” option so your PR does not pollute the PR list before you are sure that CI checks pass
+8. Wait for github to run CI, this will take about 1 hour. Make sure CI passes, and your PR will be ready for final review
+   1. Find some 2-3 nice reviewers to go over your code. If you do not know whom to choose as reviewer, ask around on discord.
+   2. Remeber, reviewers are humans and have feelings too.
+   3. If CI does not pass, fix whatever is necessary and push again into the `fix_all_bugs` branch
+10. Gotchas:
+   1. If any of the Cargo.lock files are changed by the steps above, make sure to commit them too, else CI will punish you with inscrutable errors
+   2. In some cases it may be necessary to manually [coerce Cargo.lock updates](#coerce-cargo.lock-updates)
+
+
+# Tips and Tricks
+
+## Unsorted deps {#unsorted-deps}
+
+  If your Cargo.toml is not perfectly following dtonlay's recommendations, CI will not approve your commit. Use cargo-sort to fix that.
```

Comment:
Missing a reference for these dtonlay's recommendations.

Reply:
cargo-sort is dtonlay's crate that implements them. Not sure we need another reference=)